### ISSUE TITLE
Article UID can be None

### ISSUE THREAD

**Issue by Stannislav** [BlueBrain/Search#507]
Our `bbs_database parse` takes the article UID as the output filename: [relevant line in our code](https://github.com/BlueBrain/Search/blob/8e0202d08993d961e73b39229e83c5aa1f1ee001/src/bluesearch/entrypoint/database/parse.py)

However, there are cases where the article UID can be `None`. Here's an example from `arXiv` where this is the case:
```python
import requests

from bluesearch.database.article import Article, TEIXMLParser
from bluesearch.database.pdf import grobid_pdf_to_tei_xml


def main():
    pdf_url = "https://arxiv.org/pdf/1603.01360.pdf"
    grobid_host = "dgx1.bbp.epfl.ch"
    grobid_port = 8070

    response = requests.get(pdf_url)
    tei_xml = grobid_pdf_to_tei_xml(response.content, grobid_host, grobid_port)
    parser = TEIXMLParser(tei_xml)
    article = Article.parse(parser)
    print("Article UID:", article.uid)
    print("TEI IDs in article:", parser.tei_ids)


if __name__ == "__main__":
    main()
```
Output
```
Article UID: None
TEI IDs in article: {'MD5': '034E2A6D13AD37C00D813B0E4DBEB02D'}
```

Observations
* The output filename for this article would be `None.json`
* Currently we compute `article.uid` as `generate_uid((self.pubmed_id, self.pmc_id, self.doi))`
* In this concrete example we don't even have the arXiv ID in the XML (see TEI IDs above)
* It would be good to have a UID that is not `None`.

**Comment by FrancescoCasalegno** [comment]
The "unique ID" `uid` is not so "unique" if it can be `None`.
And if we want to use it as a `PRIMARY KEY`, this requires a `UNIQUE NOT NULL` constraint.

I opened https://github.com/BlueBrain/Search/pull/538 to raise `ValueError` if all provided identifiers are `None`.

**Comment by FrancescoCasalegno** [comment]
Now, are there any other cases (other than the one provided in https://github.com/BlueBrain/Search/issues/507#issue-1061291473, which should be fixed by #534) in which all identifiers could be `None`?

#### Papers from online archives
Currently we are downloading papers from the following sources.
- `PubMed`
  - All papers have a non-null `PMID`.
- `PMC` 
  - All papers have a non-null `PMCID`.
- `arXiv` 
  - All papers have a non-null `arXiv ID`.
- `bioRxiv` 
  - ?
- `medRxiv` 
  - ?
  
 #### Other papers
 ?

**Comment by FrancescoCasalegno** [comment]
From https://github.com/BlueBrain/Search/pull/534#discussion_r778170537 by @Stannislav :
 
> Regarding potential Nones, I think according to #507 the sources we currently have are all covered.
> 
> As we discussed, we can think about a fall-back strategy for external/random texts that always works, like hashing the article title/abstract or something like that. I guess this is one of the potential places where this could be implemented. Something like
> 
> ```python
>   @property
>   def uid(self) -> str:
>       try:
>           uid = generate_uid((self.pubmed_id, self.pmc_id, self.arxiv_id, self.doi))
>       raise ValueError:
>           uid = hash_article(self)
>       return uid
>   ```
